### PR TITLE
support older PostGreSQL versions

### PR DESCRIPTION
**Pull request by zeuner**
The PostGreSQL backend support on the master branch currently requires PostGreSQL 9.5 or later. This is because the newer PostGreSQL versions support more efficient UPSERT commands. However, it may hinder adoption of the backend on OS distributions which ship an older version of PostGreSQL.

So, I suggest making the point about UPSERT efficiency a warning only, and to fall back to SQL commands which are supported on older versions if necessary. This is what this branch implements.

### CODE REVIEW

**Comment by nerzhul** [comment]
Please note your PR is wrong, you exec and insert, which can fail, in that case this will trigger an exception and do not execute the update.

**Comment by zeuner** [comment]
@nerzhul:

Thanks for the quick review.

I was hoping to minimize the amount of database writes, and to maintain some trigger compatibility to the PostGreSQL 9.5 commands. So, can you elaborate on what kind of failures you're expecting from the INSERT command? It only inserts after checking whether the primary key is already present, so the only possible failure I see is when there is a race condition between two writes to the same primary key. But this would also cause an exception with the DELETE...;INSERT... sequence.

I consider using transactions or to use client-side conditionals in order to prevent failures while avoiding DELETE...;INSERT... - so it would be of great help to better understand what kind of failures are expected.

**Comment by nerzhul** [comment]
you have two choices to solve this problem in postgresql 9.4 and lesser:

* select + insert or update depend on the row existence
* begin, delete, insert, commit

**Comment by zeuner** [comment]
Two cases to consider:

(1) Race conditions are possible

In this case, the two choices suggested won't work, either:

"select + insert or update depend on the row existence":

Thread A: SELECT ==> row not found
Thread B: SELECT ==> row not found
Thread A: INSERT ==> row exists
Thread B: INSERT ==> failure  (duplicate primary key)

"begin, delete, insert, commit":

Won't work anyway, because saveBlock might be called between calls of beginSave and endSave, resulting in nested transactions (which PostGreSQL doesn't support).

Without the transaction, it will still fail in case of race conditions:

Thread A: DELETE ==> row doesn't exist (anymore)
Thread B: DELETE ==> row still doesn't exist
Thread A: INSERT ==> row exists (again)
Thread B: INSERT ==> failure (duplicate primary key)

(2) Race conditions are made impossible by the application code

In this case, the existing implementation won't fail, so no reason to change it.

Bottom line:

The naive UPSERT approach I implemented will either work as intended, or fail with an exception. The latter case will only happen if there are race conditions where multiple threads will attempt to modify the same map position in the database. Currently, I would assume that these cases may be application code bugs anyway.

Correct me if I'm wrong and there are valid cases where the application will cause race conditions on the database writes. Until then, I'm confident that the suggested change works (it does on my test system), and that in the worst case, it may help finding application code bugs (unintended race conditions).

**Comment by nerzhul** [comment]
@zeuner there is no connection pool in our code.

creation is done by mapgen in case of the block doesn't exists. Mapblocks are saved from server thread only after made available by emerge threads. Using delete + insert using the beginsave/endsave functions make sense as this code part is called when emergethread has made mapblock available to server and mapblock was modified. Emerge can do begin/endsave too without problem as server doesn't know the mapblock at this moment

**Comment by zeuner** [comment]
@nerzhul: Since there is only one database connection in the server process, there is no way the INSERT command could fail (unless another process would write to the database during server run time, but I guess this would not be intended usage...).

Anyway, please have a look at the latest implementation. I reordered the queries so the INSERT is executed after the UPDATE. This way, I guess your concerns should be alleviated since theoretical INSERT failures cannot affect the UPDATE. Plus, the new implementation should be faster since the UPDATE queries will affect less rows.

Also, thanks for the explanation on block creation. I will keep this in mind when doing further tests. Until now, the code seems to run stable using an old PostgreSQL server, though.

By the way, is there a reason why map-mutating methods like Map::saveBlock or Map::deleteblock are defined as public methods even though only Map uses them? I guess it would be even easier to determine possible issues if the classes used stricter encapsulation.

**Comment by nerzhul** [comment]
Insertion always trigger an exception, please ensure you catch it.

Else maybe it could be good to use a function like this: http://robbat2.livejournal.com/214267.html

**Comment by zeuner** [comment]
@nerzhul: Can you elaborate on how you had it trigger an exception? It _never_ happened on my server for days. It is also not to be expected since the "WHERE NOT EXISTS" clause catches the case where the row already exists.

**Comment by nerzhul** [comment]
@zeuner sorry i misread the query, it was ok !

**Comment by nerzhul** [comment]
merged thanks;

this has been approved by @sfan5 on IRC
